src/detect-quic-cyu-string.c

```diff
@@ -0,0 +1,197 @@
+/* Copyright (C) 2020 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ *
+ * Implements the quic.cyu.string sticky buffer
+ */
+
+#include "suricata-common.h"
+#include "detect.h"
+#include "detect-parse.h"
+#include "detect-engine.h"
+#include "detect-engine-mpm.h"
+#include "detect-engine-prefilter.h"
+#include "detect-engine-content-inspection.h"
+#include "detect-quic-cyu-string.h"
+#include "rust.h"
+
+#define KEYWORD_NAME "quic.cyu.string"
+#define KEYWORD_DOC  "quic-cyu.html#quic-cyu-string"
+#define BUFFER_NAME  "quic.cyu.string"
+#define BUFFER_DESC  "QUIC CYU String"
+static int g_buffer_id = 0;
+
+struct QuicStringGetDataArgs {
+    int local_id; /**< used as index into thread inspect array */
+    void *txv;
+};
+
+static int DetectQuicCyuStringSetup(DetectEngineCtx *de_ctx, Signature *s, const char *arg)
+{
+    if (DetectBufferSetActiveList(s, g_buffer_id) < 0)
+        return -1;
+
+    if (DetectSignatureSetAppProto(s, ALPROTO_QUIC) < 0)
+        return -1;
+
+    return 0;
+}
+
+static InspectionBuffer *QuicStringGetData(DetectEngineThreadCtx *det_ctx,
+        const DetectEngineTransforms *transforms, Flow *f, struct QuicStringGetDataArgs *cbdata,
+        int list_id, bool first)
+{
+    SCEnter();
+
+    InspectionBufferMultipleForList *fb = InspectionBufferGetMulti(det_ctx, list_id);
+    InspectionBuffer *buffer = InspectionBufferMultipleForListGet(fb, cbdata->local_id);
+    if (buffer == NULL)
+        return NULL;
+    if (!first && buffer->inspect != NULL)
+        return buffer;
+
+    const uint8_t *data;
+    uint32_t data_len;
+    if (rs_quic_tx_get_cyu_string(cbdata->txv, (uint16_t)cbdata->local_id, &data, &data_len) == 0) {
```

Comment:
Based this off the an existing detect module (mqtt), should I look into doing something else?

Reply:
MQTT looks indeed to have the same problem.

You can make a better error check in `QuicStringGetData`, that would be `local_id <= UINT16_MAX`
That can be `DEBUG_VALIDATE_BUG_ON(local_id <= UINT16_MAX)` if this is meant to never happen (having more than 65536 cyu_string)

Reply:
@satta could an infinite loop happen in `DetectEngineInspectMQTTUnsubscribeTopic` ? because of this integer cast

Reply:
For MQTT, I also just took this part of the code from the DNS detection (https://github.com/OISF/suricata/blob/master/src/detect-dns-query.c#L90) which uses it in exactly the same way. So if this is a problem, it might be a pattern in more places in the Suricata code:
```
$ git grep uint16_t | grep local_id
src/detect-dns-query.c:    if (rs_dns_tx_get_query_name(cbdata->txv, (uint16_t)cbdata->local_id,
src/detect-ike-vendor.c:    if (rs_ike_tx_get_vendor(cbdata->txv, (uint16_t)cbdata->local_id, &data, &data_len) == 0) {
src/detect-krb5-cname.c:    if (rs_krb5_tx_get_cname(cbdata->txv, (uint16_t)cbdata->local_id, &b, &b_len) != 1)
src/detect-krb5-sname.c:    if (rs_krb5_tx_get_sname(cbdata->txv, (uint16_t)cbdata->local_id, &b, &b_len) != 1)
src/detect-mqtt-subscribe-topic.c:    if (rs_mqtt_tx_get_subscribe_topic(cbdata->txv, (uint16_t)cbdata->local_id,
src/detect-mqtt-unsubscribe-topic.c:    if (rs_mqtt_tx_get_unsubscribe_topic(cbdata->txv, (uint16_t)cbdata->local_id,
```
I do not see much potential for an infinite loop because the loop in the Rust code that is called there (`rs_mqtt_tx_get_unsubscribe_topic()`) iterates over a list of messages in the tx that is independent of this value.

Reply:
I think it is kind of correct for DNS as its maximum PDU length is 65536, which is not the case for MQTT (whose suricata default limit is 1Mbyte)

Reply:
@satta what would be a reasonable limit for mqtt topics ?

Reply:
This would be limited by the size of a MQTT control packet (a "message"), which can contain an arbitrary number of topics limited by the size of the message. The length is constrained by a length field in the header (https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901024) which is effectively 28 bit, so one message can be 256MB in size. 
Given that single subscribed topic in a SUBSCRIBE request needs at least 4 bytes, and the fixed amount of bytes for headers etc, and assuming that we are not using message properties, that would leave us with 67108861 topics that could be in a single subscription.

Reply:
Thanks @satta for this maximum but this is not exactly what I was asking.

Should we add a new configuration option to limit the number of topics to inspect ? And what would be its default value ?

Reply:
I guess we could, All I was saying above was that a `uint32` would be enough to fit the maximum number of potential topics. So just using that would be safe from that perspective, unless you could think of a reason not to.

Maybe for performance reasons, in practice one could indeed limit the total number of topics to inspect by setting an upper limit of loops instead of doing the `while(1)` in `DetectEngineInspectMQTT(Unsubscribe|Subscribe)Topic()`. Should we use 100 as a default? I must admit I have NO idea how many topics are typically used in actual clients.